src/linear_algebra/gram_schmidt_ortho.lean

```diff
@@ -0,0 +1,309 @@
+    have h₁ : gram_schmidt_process 𝕜 f '' Iic c.succ ⊆ span 𝕜 (f '' Iic c.succ) := h₀,
+    rwa ← span_le at h₁ },
+  { have h₀ : ∀ x ∈ f '' Iic c.succ,
+      x ∈ span 𝕜 (gram_schmidt_process 𝕜 f '' Iic c.succ),
+    { intros x hx,
+      simp only [mem_image, mem_Iic] at hx,
+      rcases hx with ⟨y, hy₁, hy₂⟩,
+      rw nat.succ_eq_add_one at hy₁ ⊢,
+      by_cases y ≤ c,
+      { have hx : x ∈ f '' Iic c := ⟨y, h, hy₂⟩,
+        have hx₁ : x ∈ span 𝕜 (f '' Iic c),
+        { have h₁ : f '' Iic c ⊆ span 𝕜 (f '' Iic c) := subset_span,
+          exact h₁ hx },
+        rw ← hc at hx₁,
+        have h₂ : span 𝕜 (gram_schmidt_process 𝕜 f '' Iic c)
+          ≤ span 𝕜 (gram_schmidt_process 𝕜 f '' Iic (c + 1)),
+        { have h₃ : gram_schmidt_process 𝕜 f '' Iic c ⊆ gram_schmidt_process 𝕜 f '' Iic (c + 1),
+          { have h₄ : Iic c ⊆ Iic (c + 1),
+            { simp only [Iic, set_of_subset_set_of],
+              intros a ha,
+              linarith },
+            exact image_subset (gram_schmidt_process 𝕜 f) h₄ },
+          exact span_mono h₃ },
+        exact h₂ hx₁ },
+      { have hy₃ : y = c + 1 := by linarith,
+        rw hy₃ at hy₂,
+        have hx : x = gram_schmidt_process 𝕜 f (c + 1) + ∑ i in finset.range (c + 1),
+          orthogonal_projection (𝕜 ∙ gram_schmidt_process 𝕜 f i) x :=
+            by simp only [gram_schmidt_process_def, hy₂, sub_add_cancel],
+        simp only [orthogonal_projection_singleton] at hx,
+        rw hx,
+        apply add_mem _ _ _,
+        { have h₁ : gram_schmidt_process 𝕜 f (c + 1) ∈ gram_schmidt_process 𝕜 f '' Iic c.succ,
+          { simp only [mem_image, mem_Iic],
+            refine ⟨c + 1, by simp only [nat.succ_eq_add_one], by refl⟩ },
+          have h₂ : gram_schmidt_process 𝕜 f '' Iic c.succ
+            ⊆ span 𝕜 (gram_schmidt_process 𝕜 f '' Iic c.succ) := subset_span,
+          exact h₂ h₁ },
+        { apply sum_mem _ _,
+          intros c₁ hc,
+          have hc₁ : gram_schmidt_process 𝕜 f c₁ ∈
+            span 𝕜 (gram_schmidt_process 𝕜 f '' Iic c.succ),
+          { have hc₂ : gram_schmidt_process 𝕜 f c₁ ∈ gram_schmidt_process 𝕜 f '' Iic c.succ,
+            { simp only [mem_image, mem_Iic, nat.succ_eq_add_one],
+              refine ⟨c₁, _, by refl⟩,
+              rw finset.mem_range at hc,
+              linarith },
+            have hc₃ : gram_schmidt_process 𝕜 f '' Iic c.succ ⊆
+              span 𝕜 (gram_schmidt_process 𝕜 f '' Iic c.succ) := subset_span,
+            exact hc₃ hc₂ },
+          exact smul_mem (span 𝕜 (gram_schmidt_process 𝕜 f '' Iic (c + 1)))
+            ((⟪gram_schmidt_process 𝕜 f c₁, x⟫ / ∥gram_schmidt_process 𝕜 f c₁∥ ^ 2) : 𝕜)
+              hc₁ }}},
+    have h₁ : f '' Iic c.succ ⊆ span 𝕜 (gram_schmidt_process 𝕜 f '' Iic c.succ) := h₀,
+    rwa ← span_le at h₁ }
+end
+
+/-- If the input of gram_schmidt_process is linearly independent, then output is non-zero -/
+lemma gram_schmidt_process_ne_zero (f : ℕ → E) (h₀ : linear_independent 𝕜 f) (n : ℕ) :
```

Comment:
I think this is too strong, only the first n need to be linearly independent.

Reply:
I was thinking that, but I am not sure what would be more appropriate here? May you give me some suggestions here?

Reply:
I was also thinking that, if I want to prove orthonormalization in my code, I still need to use this 'strong' lemma here
``` theorem gram_schmidt_process_orthonormal (f : ℕ → E) (h₀ : linear_independent 𝕜 f) :
  orthonormal 𝕜 (gram_schmidt_process_normed 𝕜 f) := ```

Reply:
I mean something like
```suggestion
lemma gram_schmidt_process_ne_zero (f : ℕ → E) (n : ℕ) (h₀ : linear_independent 𝕜 (f ∘ (coe : fin n → ℕ)) :
```

Reply:
Oh, I see. May I keep this one and write another version like this? Since I think I still need this lemma in my proof of orthonormalization

Reply:
The version you have follows trivially from the one I'm suggesting.

Reply:
`linear_independent.comp` or something

Reply:
Also, I think I meant `(coe : fin n.succ → ℕ)`

Reply:
Ok,I will have a go with it, thank you very much!